File: LeetCode_Monthly_Challenge/July/Python3/3Sum.py
"""
PROBLEM:
Given an array nums of n integers, are there elements a, b, c in nums such that a + b + c = 0? Find all unique triplets in the array which gives the sum of zero.
Note:
The solution set must not contain duplicate triplets.
Example:
Given array nums = [-1, 0, 1, 2, -1, -4],
A solution set is:
[
  [-1, 0, 1],
  [-1, -1, 2]
]
"""



class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums=sorted(nums)
        result = []
        for i in range(len(nums)):
            if nums[i] > 0: 
                continue
            j = i + 1 
            k = len(nums) - 1
            while j < k:
                sum = nums[i] + nums[j] + nums[k]
                if sum < 0:
                    j+=1
                elif sum > 0: 
                    k-=1
                else:
                    result.append([nums[i], nums[j], nums[k]])
                    j+=1
                    k-=1
        return result

File: LeetCode_Monthly_Challenge/July/C++/Week_2/3Sum.cpp
/*
PROBLEM:
Given an array nums of n integers, are there elements a, b, c in nums such that a + b + c = 0? Find all unique triplets in the array which gives the sum of zero.
Note:
The solution set must not contain duplicate triplets.
Example:
Given array nums = [-1, 0, 1, 2, -1, -4],
A solution set is:
[
  [-1, 0, 1],
  [-1, -1, 2]
]
*/



class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(),nums.end());
        vector<vector<int>> result;
        for (int i = 0; i < nums.size(); i++) {
            if (nums[i] > 0) 
                continue;
            int j = i + 1, k = nums.size() - 1;
            while (j < k) {
                int sum = nums[i] + nums[j] + nums[k];
                if (sum < 0) j++;
                else if (sum > 0) k--;
                else {
                    result.push_back({nums[i], nums[j], nums[k]});
                    // Skip duplicate numbers of j
                    while (j+1 <= k && nums[j] == nums[j+1]) j++;
                    j++;
                    k--;
                }
            }
            // Skip duplicate numbers of i
            while (i + 1 < nums.size() && nums[i + 1] == nums[i]) i++;
        }
        return result;
    }
};
